lib/internal/readline/eventsToAsyncIteratorFactory.js

```diff
@@ -0,0 +1,107 @@
+'use strict';
+const {
+  Promise,
+  SymbolAsyncIterator,
+  ArrayPrototypeConcat,
+} = primordials;
+const FixedQueue = require('internal/fixed_queue');
+
+const PAUSE_THRESHOLD = 1024;
+const RESUME_THRESHOLD = 1;
+const ITEM_EVENTS = ['data'];
+const CLOSE_EVENTS = ['close', 'end'];
+const ERROR_EVENTS = ['error'];
+
+
+function waitNext(emitter, next, events) {
+  return new Promise((resolve, reject) => {
```

Comment:
Can you explain why this isn't a `Promise.all(events.map(name => once(emitter, name))`? 

Once is really nice (and converts a single time event to a promise) - the error handling would be dealt with automatically too I think.

If there is a good reason not to use the helper - perhaps `emitter.once` would still be better sugar?

Reply:
Also, note there is an async iterator version for it `events.on` which can be useful

Reply:
> Can you explain why this isn't a Promise.all(events.map(name => once(emitter, name))?
Notice that my promise can be resolved on the line, error, and close event, basically. During the 'line' event, the error would be treated automatically, but not the close, so, if I have a very slow stream, this promise needs to be created many times and I'd accumulate a lot of listeners on the close event, resulting in an event leak. I could write some code that uses the once function, but as I'd need to remove some listeners manually, it'd end up being as complex and less light than the current code.
Maybe it's good to write a unit test to validate this specific scenario of slow stream, I think.

> If there is a good reason not to use the helper - perhaps emitter.once would still be better sugar?
under the hood .once executes an addListener passing a wrapped function that removes the listener. As I have N events that can resolve the promise and I need to remove the listener of all of them, it's preferable not to use once and just always remove all the events in the callback

Reply:
> Also, note there is an async iterator version for it `events.on` which can be useful

about that I'll take a look to understand better, I didn't know about that